cloudapi/logs.go

```diff
@@ -164,8 +199,25 @@ func (c *Config) StreamLogsToLogger(
 		}
 
 		if err != nil {
-			logger.WithError(err).Warn("error reading a message from the cloud")
+			logger.WithError(err).Warn("error reading a log message from the cloud, trying to establish a fresh connection with the logs service...") //nolint:lll
+
+			// TODO: avoid the "logical" race condition
+			// The case explained:
+			// * The msgBuffer consumer is slow
+			// * ReadMessage is fast and adds at least one more message in the buffer
+			// * An error is got in the meantime and the re-dialing procedure is tried
+			// * Then the latest timestamp used will not be the real latest received
+			// * because it is still waiting to be processed.
+			// In the case the connection will be restored then the first message will be a duplicate.
+			newconn, errd := c.logtailConn(ctx, referenceID, latest.TimeOrNow())
+			if errd == nil {
+				mconn.Lock()
+				conn = newconn
+				mconn.Unlock()
+				continue
+			}
 
+			// return the main error
 			return err
```

Comment:
nit: I would've done `errd!=nil` return `err` and than skip the `continue` but also keep the good case with smaller identation

Reply:
>then skip the continue

I reversed the check having the classic `!= nil` but I'm not sure I got how you would change to avoid the continue.